### PR TITLE
Generate JNI fuction following JNI standards

### PR DESCRIPTION
**Pull request by bowemi**
When java package names contain `_`, the underscore needs to be escaped
as `_1`. Without this escaping, Java can not find the native function

This change also proposes a test framework for Java initiated
applications calling into Rust generated functions.

fixes https://github.com/duchess-rs/duchess/issues/166

### CODE REVIEW

File: macro/src/class_info.rs [hunk 1]
```diff
@@ -708,10 +708,27 @@ impl DotId {
...
+        self.split().1.data.replace("_", "_1").into()
+    }
+
     pub fn class_name(&self) -> &Id {
         self.split().1
     }
 
+    /// returns the package in JNI format with _'s escaped with _1
+    /// https://docs.oracle.com/en/java/javase/17/docs/specs/jni/design.html
+    pub fn to_jni_packge(&self) -> String {
```
Comment (nikomatsakis) [inline macro/src/class_info.rs:723]:
```suggestion
    pub fn to_jni_package(&self) -> String {
```

File: test-crates/duchess-java-tests/tests/ui/native_fn_callable_from_java.rs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+//@compile-flags: --crate-type cdylib
+//@check-pass
+//@rustc-env: CLASSPATH=../target/tests/java_ui
```
Comment (nikomatsakis) [inline test-crates/duchess-java-tests/tests/ui/native_fn_callable_from_java.rs:3]: Hmm, I like this capability, I don't know if I like mixing it in with the other `ui` tests. What about making a separate directory where these compile-flags and rustc-env options are the default? Something like:

* `src/test/rs-to-java`
* `src/test/rs-from-java` <-- the new one
* `src/test/js-to-rs`

or even just put the `.rs` files into the `src/test/js-to-rs` directory as well?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
What do you think?

File: test-crates/duchess-java-tests/tests/ui/native_fn_callable_from_java.rs (hunk 2 above)
Comment (bowemi) [reply test-crates/duchess-java-tests/tests/ui/native_fn_callable_from_java.rs:3]: I like your 3 directory proposal, I'll go with that, but as far as naming and structure, what about...

* ```src/test/java-to-rust/java```
* ```src/test/java-to-rust/rust_libraries```
* ```src/test/rs-to-java```

**Review by bowemi — COMMENTED** [review]
Moved the ui test files into the directory structure based on Niko's comment, also correct the function ```to_jni_packge``` -> ```to_jni_package```

**Review by nikomatsakis — APPROVED** [review]
Very nice.

**Comment by nikomatsakis** [comment]
> This branch has conflicts that must be resolved

unfortunately, can't merge yet :/

**Pull request merged**